File: mobile/shared/src/commonTest/kotlin/uk/foodclassifier/image/NormalizedRectTest.kt
package uk.foodclassifier.image

import kotlin.test.Test
import kotlin.test.assertEquals

class NormalizedRectTest {
    @Test
    fun translatedClampsToImageBounds() {
        val rect = NormalizedRect(0.1f, 0.2f, 0.5f, 0.7f)
        val moved = rect.translated(normalizedDeltaX = 0.8f, normalizedDeltaY = 0.8f)
        assertEquals(1.0f, moved.right)
        assertEquals(1.0f, moved.bottom)
    }

    @Test
    fun fitImageBoundsLetterboxesWideImage() {
        val bounds = fitImageBounds(
            containerWidth = 100f,
            containerHeight = 200f,
            imageWidth = 200f,
            imageHeight = 100f,
        )
        assertEquals(75f, bounds.top)
        assertEquals(50f, bounds.height)
    }

    @Test
    fun mapViewGuideToImageGuideAccountsForLetterboxing() {
        val guide = NormalizedRect(0.04f, 0.36f, 0.96f, 0.58f)
        val mapped = mapViewGuideToImageGuide(
            guide = guide,
            viewWidth = 100f,
            viewHeight = 200f,
            imageWidth = 200f,
            imageHeight = 100f,
        )
        assertEquals(0.04f, mapped.left)
        assertEquals(0.0f, mapped.top)
        assertEquals(0.96f, mapped.right)
        assertEquals(0.82f, mapped.bottom)
    }
}

File: flutter/test/normalized_rect_test.dart
import 'package:flutter_test/flutter_test.dart';
import 'package:food_classifier/image/normalized_rect.dart';

void main() {
  test('translatedClampsToImageBounds', () {
    const rect = NormalizedRect(left: 0.1, top: 0.2, right: 0.5, bottom: 0.7);
    final moved = rect.translated(normalizedDeltaX: 0.8, normalizedDeltaY: 0.8);
    expect(moved.right, 1.0);
    expect(moved.bottom, 1.0);
  });

  test('fitImageBoundsLetterboxesWideImage', () {
    final bounds = fitImageBounds(
      containerWidth: 100,
      containerHeight: 200,
      imageWidth: 200,
      imageHeight: 100,
    );
    expect(bounds.top, 75);
    expect(bounds.height, 50);
  });

  test('mapViewGuideToImageGuideAccountsForLetterboxing', () {
    const guide = NormalizedRect.ingredientsGuide;
    final mapped = mapViewGuideToImageGuide(
      guide: guide,
      viewWidth: 100,
      viewHeight: 200,
      imageWidth: 200,
      imageHeight: 100,
    );
    expect(mapped.left, 0.04);
    expect(mapped.top, 0.0);
    expect(mapped.right, 0.96);
    expect(mapped.bottom, closeTo(0.82, 0.001));
  });
}
